### PR TITLE
Bignum API

### PR DESCRIPTION
**Pull request by polubelova**
This PR introduces the following
- Bignum API: [32-bit](https://github.com/project-everest/hacl-star/blob/139fc5caf6b24f73f955c919754d3b266c73712d/dist/gcc-compatible/Hacl_Bignum32.h) and [64-bit](https://github.com/project-everest/hacl-star/blob/139fc5caf6b24f73f955c919754d3b266c73712d/dist/gcc-compatible/Hacl_Bignum64.h)
- Montgomery API (field arithmetic): [32-bit](https://github.com/project-everest/hacl-star/blob/139fc5caf6b24f73f955c919754d3b266c73712d/dist/gcc-compatible/Hacl_GenericField32.h) and [64-bit](https://github.com/project-everest/hacl-star/blob/139fc5caf6b24f73f955c919754d3b266c73712d/dist/gcc-compatible/Hacl_GenericField64.h)
- Almost Montgomery Multiplication
- Double exponentiation

### CODE REVIEW

File: code/bignum/Hacl.Bignum.SafeAPI.fst [hunk 1]
```diff
@@ -13,19 +13,20 @@ module B = LowStar.Buffer
 module HS = FStar.HyperStack
 module LSeq = Lib.Sequence
 
+module E = Hacl.Spec.Exponentiation.Lemmas
```
Comment (beurdouche) [inline code/bignum/Hacl.Bignum.SafeAPI.fst:16]: `SEL`?

File: code/bignum/Makefile [hunk 2]
```diff
@@ -7,15 +7,16 @@ all: dist/bignum-test.exe
 # Defines rules for producing .checked, .krml, .depend, etc.
 include ../../Makefile.local
 
-CFLAGS += -I ../../../lib/c -I ../../lib/c -march=native -mtune=native -O3
+CFLAGS += -I ../../../lib/c -I ../../lib/c -march=native -mtune=native -O3 -DCOMPILE_INTRINSICS
```
Comment (beurdouche) [inline code/bignum/Makefile:10]: Is this really needed? If yes, does someone need to update all the build systems for all products currently using HACL* ?

**Review by beurdouche — APPROVED** [review]
I went through this pretty thoroughly and it looks very nice.
I generally like the naming conventions much better. There are still places in the internal functions where suffixes like `_f_` or `_r_` could be improved a bit, but overall this is great work !! : )

File: code/bignum/Makefile (hunk 2 above)
Comment (polubelova) [reply code/bignum/Makefile:10]: As long as they run the configure script, no update is required.
We don't run ./configure for local makefiles, that's why this change is needed.
[This](https://github.com/project-everest/hacl-star/commit/57c68278a4574038956f5b64b53ca8dd42abbd50) commit has changed the default behavior.

Comment (beurdouche) [reply code/bignum/Makefile:10]: Nobody runs the configure file.

Comment (beurdouche) [reply code/bignum/Makefile:10]: So the question is: what will break if this flag is not passed?

Comment (polubelova) [reply code/bignum/Makefile:10]: It will take a fallback implementation or complain that the Hacl_IntTypes_Intrinsics.c file is not found.

**Review by prosecco — APPROVED** [review]

**Pull request merged**